### PR TITLE
ROU-3054: Style client actions should return error message and success boolean

### PR DESCRIPTION
**Pull request by glundgren93**
This PR adds feedback messages to Style API methods. All methods did not return any feedback on whether or not the operation was successful or if any errors happened. Also, fixes an [issue reported on this forum post](https://www.outsystems.com/forums/discussion/73148/add-and-remove-css-classes/)

### Test Steps

1. Click in "AddCellCssClass":
Expected: The first cell from ProductName should become red (you might need to click on another cell) and a success message should appear.
2. Click in "RemoveAllCssClassesFromCell":
Expected: The first cell from ProductName should have the default color and a success message should appear
 
3. Click in "AddColumnCssClass":
Expected: The ProductName column should become red and a success message should appear.
4. Click in "RemoveCssClassesFromColumn":
Expected: The ProductName column should have the default color and a success message should appear.

5. Click in "AddClass":
Expected: The first row column should become red and a success message should appear.
6. Click in "RemoveClass":
Expected: The first row should have the default color and a success message should appear.

7. Click in "AddClass":
8. Click in "Remove All Classes From Row ":
Expected: The first row should have the default color and a success message should appear.

1. Click in "AddCellCssClass invalid":
Expected: Error message should show saying: "Error"

2. Click in "RemoveAllCssClassesFromCell  invalid":
Expected: Error message should show saying: "Error"
 
3. Click in "AddColumnCssClass  invalid":
Expected: Error message should show saying: "Error"

4. Click in "RemoveCssClassesFromColumn  invalid":
Expected: Error message should show saying: "Error"

5. Click in "AddClass invalid" :
Expected: Error message should show saying: "Error"

6. Click in "RemoveClass  invalid":
Expected: Error message should show saying: "Error"

7. Click in "Remove All Classes From Row  invalid":
Expected: Error message should show saying: "Error"

### CODE REVIEW

**Review by BenOsodrac — DISMISSED** [review]

**Review by joselrio — DISMISSED** [review]

File: code/src/GridAPI/Rows.ts [hunk 1]
```diff
@@ -15,20 +15,44 @@ namespace GridAPI.Rows {
...
         rowNumber: number,
         className: string
-    ): void {
+    ): string {
         PerformanceAPI.SetMark('Rows.AddClass');
+        let returnMessage = {
+            isSuccess: true,
+            message: 'Success',
+            code: OSFramework.Enum.ErrorCodes.GRID_SUCCESS
+        };
```
Comment (rugoncalves) [inline code/src/GridAPI/Rows.ts:24]: Please have an action, to be used cross API to do the set of this default object. Repeated with https://github.com/OutSystems/outsystems-datagrid-reactive/pull/188

File: code/src/GridAPI/Rows.ts [hunk 2]
```diff
@@ -15,20 +15,44 @@ namespace GridAPI.Rows {
...
 
-        const grid = GridManager.GetGridById(gridID);
-
-        if (grid !== undefined) {
-            grid.features.rows.addClass(rowNumber, className, true);
+        if (!OSFramework.Helper.IsGridReady(gridID)) {
+            returnMessage = {
+                isSuccess: false,
+                message: 'Grid not found',
```
Comment (rugoncalves) [inline code/src/GridAPI/Rows.ts:29]: This text should not be hardcoded here. We should have an Enum ErrorMessages. Repeated with https://github.com/OutSystems/outsystems-datagrid-reactive/pull/188

File: code/src/GridAPI/Rows.ts [hunk 3]
```diff
@@ -15,20 +15,44 @@ namespace GridAPI.Rows {
...
+                rowNumber,
+                className,
+                true
+            );
+        } catch (error) {
+            returnMessage = {
+                isSuccess: false,
+                message: 'Error',
+                code: OSFramework.Enum.ErrorCodes.API_FailedAddClass
+            };
```
Comment (rugoncalves) [inline code/src/GridAPI/Rows.ts:45]: here you can access to the properties of the type, e.g. `returnMessage.isSuccess = false`.
Also please change the name of the object to something different from "returnMessage" :) 
Repeated with https://github.com/OutSystems/outsystems-datagrid-reactive/pull/188

File: code/src/GridAPI/Rows.ts [hunk 4]
```diff
@@ -106,20 +130,45 @@ namespace GridAPI.Rows {
...
+    export function RemoveAllClasses(
+        gridID: string,
+        rowNumber: number
+    ): string {
         PerformanceAPI.SetMark('Rows.RemoveAllClasses');
+        let returnMessage = {
+            isSuccess: true,
+            message: 'Success',
+            code: OSFramework.Enum.ErrorCodes.GRID_SUCCESS
+        };
```
Comment (rugoncalves) [inline code/src/GridAPI/Rows.ts:143]: See above comments.

File: code/src/GridAPI/Rows.ts [hunk 5]
```diff
@@ -106,20 +130,45 @@ namespace GridAPI.Rows {
...
 
-        const grid = GridManager.GetGridById(gridID);
+        if (!OSFramework.Helper.IsGridReady(gridID)) {
+            returnMessage = {
+                isSuccess: false,
+                message: 'Grid not found',
```
Comment (rugoncalves) [inline code/src/GridAPI/Rows.ts:147]: See above comments.

File: code/src/GridAPI/Rows.ts [hunk 6]
```diff
@@ -106,20 +130,45 @@ namespace GridAPI.Rows {
...
+        }
 
-        if (grid !== undefined) {
+        try {
+            const grid = GridManager.GetGridById(gridID);
             grid.features.rows.removeAllClasses(rowNumber);
+        } catch (error) {
+            returnMessage = {
+                isSuccess: false,
+                message: 'Error',
```
Comment (rugoncalves) [inline code/src/GridAPI/Rows.ts:159]: See above comments.

File: code/src/GridAPI/Rows.ts [hunk 7]
```diff
@@ -134,20 +183,44 @@ namespace GridAPI.Rows {
...
         rowNumber: number,
         className: string
-    ): void {
+    ): string {
         PerformanceAPI.SetMark('Rows.RemoveClass');
+        let returnMessage = {
+            isSuccess: true,
+            message: 'Success',
+            code: OSFramework.Enum.ErrorCodes.GRID_SUCCESS
+        };
```
Comment (rugoncalves) [inline code/src/GridAPI/Rows.ts:192]: See above comments.

File: code/src/GridAPI/Rows.ts [hunk 8]
```diff
@@ -134,20 +183,44 @@ namespace GridAPI.Rows {
...
 
-        const grid = GridManager.GetGridById(gridID);
-
-        if (grid !== undefined) {
-            grid.features.rows.removeClass(rowNumber, className, true);
+        if (!OSFramework.Helper.IsGridReady(gridID)) {
+            returnMessage = {
+                isSuccess: false,
+                message: 'Grid not found',
```
Comment (rugoncalves) [inline code/src/GridAPI/Rows.ts:197]: See above comments.

File: code/src/GridAPI/Rows.ts [hunk 9]
```diff
@@ -134,20 +183,44 @@ namespace GridAPI.Rows {
...
+                rowNumber,
+                className,
+                true
+            );
+        } catch (error) {
+            returnMessage = {
+                isSuccess: false,
+                message: 'Error',
+                code: OSFramework.Enum.ErrorCodes.API_FailedRemoveClass
+            };
```
Comment (rugoncalves) [inline code/src/GridAPI/Rows.ts:213]: See above comments.

File: code/src/GridAPI/Styling.ts [hunk 10]
```diff
@@ -17,22 +17,57 @@ namespace GridAPI.Styling {
...
         rowIndex: number,
         className: string
-    ): void {
+    ): string {
         PerformanceAPI.SetMark('Styling.SetCellCssClass');
+        let returnMessage = {
+            isSuccess: true,
+            message: 'Success',
+            code: OSFramework.Enum.ErrorCodes.GRID_SUCCESS
+        };
```
Comment (rugoncalves) [inline code/src/GridAPI/Styling.ts:26]: See above comments.

File: code/src/GridAPI/Styling.ts [hunk 11]
```diff
@@ -17,22 +17,57 @@ namespace GridAPI.Styling {
...
+
+        if (!OSFramework.Helper.IsGridReady(gridID)) {
+            returnMessage = {
+                isSuccess: false,
+                message: 'Grid not found',
```
Comment (rugoncalves) [inline code/src/GridAPI/Styling.ts:31]: See above comments.

File: code/src/GridAPI/Styling.ts [hunk 12]
```diff
@@ -17,22 +17,57 @@ namespace GridAPI.Styling {
...
+                GridManager.GetGridById(gridID).features.cellStyle.addClass(
+                    binding,
+                    rowIndex,
+                    className,
+                    true
+                );
+            } else {
+                returnMessage = {
+                    isSuccess: false,
+                    message: 'It seems you are not passing a valid column.',
```
Comment (rugoncalves) [inline code/src/GridAPI/Styling.ts:51]: See above comments.

File: code/src/GridAPI/Styling.ts [hunk 13]
```diff
@@ -17,22 +17,57 @@ namespace GridAPI.Styling {
...
+                    code: OSFramework.Enum.ErrorCodes.API_FailedSetCellCssClass
+                };
+            }
+        } catch (error) {
+            returnMessage = {
+                isSuccess: false,
+                message: 'Error',
```
Comment (rugoncalves) [inline code/src/GridAPI/Styling.ts:58]: See above comments.

File: code/src/GridAPI/Styling.ts [hunk 14]
```diff
@@ -49,21 +84,45 @@ namespace GridAPI.Styling {
...
         cssClass: string,
         applyToHeader: boolean
-    ): void {
+    ): string {
         PerformanceAPI.SetMark('Styling.SetColumnCssClass');
+        let returnMessage = {
+            isSuccess: true,
+            message: 'Success',
+            code: OSFramework.Enum.ErrorCodes.GRID_SUCCESS
+        };
```
Comment (rugoncalves) [inline code/src/GridAPI/Styling.ts:93]: See above comments.

File: code/src/GridAPI/Styling.ts [hunk 15]
```diff
@@ -49,21 +84,45 @@ namespace GridAPI.Styling {
...
+
+        if (!OSFramework.Helper.IsGridReady(gridID)) {
+            returnMessage = {
+                isSuccess: false,
+                message: 'Grid not found',
```
Comment (rugoncalves) [inline code/src/GridAPI/Styling.ts:98]: See above comments.

File: code/src/GridAPI/Styling.ts [hunk 16]
```diff
@@ -49,21 +84,45 @@ namespace GridAPI.Styling {
...
+        try {
+            GridManager.GetGridById(gridID).features.styling.addColumnCssClass(
+                columnID,
+                cssClass,
+                applyToHeader
+            );
+        } catch (error) {
+            returnMessage = {
+                isSuccess: false,
+                message: 'Error',
```
Comment (rugoncalves) [inline code/src/GridAPI/Styling.ts:113]: See above comments.

File: code/src/GridAPI/Styling.ts [hunk 17]
```diff
@@ -77,24 +136,57 @@ namespace GridAPI.Styling {
         gridID: string,
         columnID: string,
         rowIndex: number
-    ): void {
+    ): string {
         PerformanceAPI.SetMark('Styling.RemoveAllCssClassesFromCell');
+        let returnMessage = {
+            isSuccess: true,
+            message: 'Success',
```
Comment (rugoncalves) [inline code/src/GridAPI/Styling.ts:143]: See above comments.

File: code/src/GridAPI/Styling.ts [hunk 18]
```diff
@@ -77,24 +136,57 @@ namespace GridAPI.Styling {
...
-        if (column !== undefined) {
-            const binding = column.config.binding;
-            GridManager.GetGridById(gridID).features.cellStyle.removeAllClasses(
-                rowIndex,
-                binding
-            );
+        if (!OSFramework.Helper.IsGridReady(gridID)) {
+            returnMessage = {
+                isSuccess: false,
+                message: 'Grid not found',
```
Comment (rugoncalves) [inline code/src/GridAPI/Styling.ts:150]: See above comments.

File: code/src/GridAPI/Styling.ts [hunk 19]
```diff
@@ -77,24 +136,57 @@ namespace GridAPI.Styling {
...
+            if (column !== undefined) {
+                const binding = column.config.binding;
+
+                GridManager.GetGridById(
+                    gridID
+                ).features.cellStyle.removeAllClasses(rowIndex, binding, true);
+            } else {
+                returnMessage = {
+                    isSuccess: false,
+                    message: 'It seems you are not passing a valid column.',
```
Comment (rugoncalves) [inline code/src/GridAPI/Styling.ts:168]: See above comments.

File: code/src/GridAPI/Styling.ts [hunk 20]
```diff
@@ -77,24 +136,57 @@ namespace GridAPI.Styling {
...
+                        .API_FailedRemoveAllCssClassesFromCell
+                };
+            }
+        } catch (error) {
+            returnMessage = {
+                isSuccess: false,
+                message: 'Error',
+                code: OSFramework.Enum.ErrorCodes
+                    .API_FailedRemoveAllCssClassesFromCell
+            };
```
Comment (rugoncalves) [inline code/src/GridAPI/Styling.ts:179]: See above comments.

File: code/src/GridAPI/Styling.ts [hunk 21]
```diff
@@ -108,20 +200,42 @@ namespace GridAPI.Styling {
...
         columnID: string,
         cssClass: string
-    ): void {
+    ): string {
         PerformanceAPI.SetMark('Styling.RemoveColumnCssClass');
+        let returnMessage = {
+            isSuccess: true,
+            message: 'Success',
+            code: OSFramework.Enum.ErrorCodes.GRID_SUCCESS
+        };
```
Comment (rugoncalves) [inline code/src/GridAPI/Styling.ts:209]: See above comments.

File: code/src/GridAPI/Styling.ts [hunk 22]
```diff
@@ -108,20 +200,42 @@ namespace GridAPI.Styling {
...
 
-        if (!OSFramework.Helper.IsGridReady(gridID)) return;
-        GridManager.GetGridById(gridID).features.styling.removeColumnCssClass(
-            columnID,
-            cssClass
-        );
+        if (!OSFramework.Helper.IsGridReady(gridID)) {
+            returnMessage = {
+                isSuccess: false,
+                message: 'Grid not found',
```
Comment (rugoncalves) [inline code/src/GridAPI/Styling.ts:214]: See above comments.

File: code/src/GridAPI/Styling.ts [hunk 23]
```diff
@@ -108,20 +200,42 @@ namespace GridAPI.Styling {
...
+        }
+
+        try {
+            GridManager.GetGridById(
+                gridID
+            ).features.styling.removeColumnCssClass(columnID, cssClass);
+        } catch (error) {
+            returnMessage = {
+                isSuccess: false,
+                message: 'Error',
```
Comment (rugoncalves) [inline code/src/GridAPI/Styling.ts:227]: See above comments.

File: code/src/WijmoProvider/Features/CellStyle.ts [hunk 24]
```diff
@@ -76,17 +82,32 @@ namespace WijmoProvider.Feature {
...
 
-        public removeAllClasses(rowNumber: number, binding: string): void {
+        public removeAllClasses(
+            rowNumber: number,
+            binding: string,
+            refresh = false
+        ): void {
             this.getMetadata(rowNumber).removeAllClasses(binding);
+
+            // we only want to force refresh on specific cases, as this could cause infinite renders on grid
```
Comment (rugoncalves) [inline code/src/WijmoProvider/Features/CellStyle.ts:92]: How did we never caught that before?

**Review by rugoncalves — CHANGES_REQUESTED** [review]

File: code/src/GridAPI/Rows.ts (hunk 1 above)
Comment (glundgren93) [reply code/src/GridAPI/Rows.ts]: done

File: code/src/GridAPI/Rows.ts (hunk 2 above)
Comment (glundgren93) [reply code/src/GridAPI/Rows.ts]: done

File: code/src/GridAPI/Rows.ts (hunk 4 above)
Comment (glundgren93) [reply code/src/GridAPI/Rows.ts]: done

File: code/src/WijmoProvider/Features/CellStyle.ts (hunk 24 above)
Comment (glundgren93) [reply code/src/WijmoProvider/Features/CellStyle.ts:92]: We had a forum post opened in September, but we never fixed it.

Comment (rugoncalves) [reply code/src/WijmoProvider/Features/CellStyle.ts:92]: Next time, separate PR :)

**Review by rugoncalves — APPROVED** [review]
Lovely!

**Review by joselrio — APPROVED** [review]

**Pull request merged**